### DISCUSSION TITLE
Missing features on Java SCORE to build PRA BMV

### DISCUSSION THREAD

**Discussion by flowersinthesand** [icon-project/btp#5] [Q&A] [answered]
We found the following features are needed on Java SCORE to build PRA BMV, and Polkadot Java Client, Polkaj [1], provides necessary implementations you can adopt.

1. ~~SCALE codec encoding/decoding support [2] to encode block header, value of Merkle Patricia Trie (MPT), and validator signature messages. See [3] for Polkaj's support.~~ ICON DAO will implement it
2. xxHash128, Blake2b 128, Blake2b 256 hashing functions [5] to verify MPT proof and block. Polkaj wraps their api using artifacts from https://www.bouncycastle.org/ for Blake2 and net.openhft:zero-allocation-hashing for xxHash.
3. Ed25519 signature scheme [6] to verify message signatures. You can see Polkajs' support at [7].

Reference

[1] https://github.com/emeraldpay/polkaj
[2] https://substrate.dev/docs/en/knowledgebase/advanced/codec
[3] https://github.com/emeraldpay/polkaj/blob/master/docs/02-encoding.adoc
[5] https://www.shawntabrizi.com/substrate/substrate-storage-deep-dive/
[6] https://substrate.dev/docs/en/knowledgebase/advanced/cryptography#ed25519
[7] https://github.com/emeraldpay/polkaj/blob/master/docs/04-cryptography.adoc#verify-message-signature

**Comment by cow-hs** [comment, marked as answer]
It is recommended to implement the SCALE codec on its own.

The remaining hashes and signatures will be added to the API.
We will notify you when the related schedule is decided.

**Reply by flowersinthesand** [reply]
> It is recommended to implement the SCALE codec on its own.

Noted.

**Reply by cow-hs** [reply]
we will plan to support 2,3 until next week end

**Reply by flowersinthesand** [reply]
@leeheonseung 

> we will plan to support 2,3 until next week end

I'd like to remind tomorrow is the end of the week.

**Reply by cow-hs** [reply]
We have just supported that in master branch.
Please pull latest master branch and build Javaee-api.jar then apply your test environment.
ref. [gochain_icon_local_node_guide.md](https://github.com/icon-project/goloop/blob/master/doc/gochain_icon_local_node_guide.md)
ref. [APITest.java/computeHash](https://github.com/icon-project/goloop/blob/a62641ac63d1ec31c097ec9801b53dcd1226c0b5/javaee/samples/testcases/src/testcases/APITest.java#L192)
ref. [Crypto.java](https://github.com/icon-project/goloop/blob/a62641ac63d1ec31c097ec9801b53dcd1226c0b5/javaee/rt/src/java/foundation/icon/ee/util/Crypto.java#L63)

**Reply by danielvo11** [reply]
I confirm that it work. Could you support `ed25519` algorithm for `recoverKey` function? I input `ed25519` but get `IllegalArgumentException`. Thank you for your support!.

**Reply by cow-hs** [reply]
@danielAlvess
we already supported about Ed25519 for verify message signatures.
you meant you need additional Ed25519 algorithm for recoverKey right?

**Reply by danielvo11** [reply]
> we already supported about Ed25519 for verify message signatures.

Yes, I know. My input is list of signatures (about 2/3* 110 + 1 = 74 signatures) and list of 110 validator addresses. I need to verify 74 signatures is valid and sign by validator addresses. If I use verify signatures function, I need additional data to determine address which use to sign. About 32 bytes per address * 74 = 2368 bytes.

If you could support `Ed25519` for recover key, I just need to extract public key from signatures and compare it with validator's address. So that I can save 2368 bytes of data input to smart contract that is large data in Smart Contract environment.

Because of that reason, it will be great if you can support Ed25519 in `recoverKey` function. Thank you!

**Reply by cow-hs** [reply]
@danielAlvess 
we consider about above comment until early next week.

**Reply by cow-hs** [reply]
@danielAlvess 
Ed25519 uses EdDSA (not ECDSA) and EdDSA cannot recoverKey.

ref.[Polkadot Keys - Polkadot Wiki](https://wiki.polkadot.network/docs/en/learn-keys#why-was-ed25519-selected-over-secp256k1)
"There are obsolete Schnorr variants that support recovering the public key from a signature, but they break important functionality like hierarchical deterministic key derivation."

**Reply by danielvo11** [reply]
Thank you for your information.  So I will implement it using `verifySignature` function.
